### PR TITLE
Download links grouped by WFS layer

### PR DESCRIPTION
**Pull request by pmbohm**
To test use Geonetwork-123 and uuid=0a21e0b9-8acb-4dc2-8c82-57c3ea94dd85

### CODE REVIEW

**Comment by pmbohm** [comment]
We need a field to be manually configured in the MEST. Without it we will end up with meaningless grouping such as this for ARGO
screen shot 2015-09-15 at 3 15 42 pm

File: src/test/javascript/portal/cart/DownloadPanelSpec.js [hunk 1]
```diff
@@ -234,6 +235,17 @@ describe("Portal.cart.DownloadPanel", function() {
         });
     });
 
+    describe('getEmbededTitle', function() {
+
+        it('Extracts the real title super trickerly embeded title in the description field from Geonetwork', function() {
```
Comment (jkburges) [inline src/test/javascript/portal/cart/DownloadPanelSpec.js]: I'm not really sure what's going on here - are there metadata records which have a `description` field, which in turn contain a "title", given in parentheses?  Is that right?  (Example metadata record, please).

This seems a bit weird to me - why not put the "title" separately in the metadata record?

File: src/test/javascript/portal/prototypes/ArraySpec.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+describe("Array", function() {
+
+    var obj;
+
+    it('hasDuplicates', function() {
+
+        obj = {test: "astringval"};
+        expect([1,2,3,4,5,4,3,3,2,1,undefined].hasDuplicates()).toEqual(true);
```
Comment (jkburges) [inline src/test/javascript/portal/prototypes/ArraySpec.js]: coding standard, `expect([1,2,3,4` should be `expect([1, 2, 3, 4` and so on.

File: web-app/js/portal/cart/DownloadPanel.js [hunk 3]
```diff
@@ -158,18 +158,37 @@ Portal.cart.DownloadPanel = Ext.extend(Ext.Panel, {
         setViewPortTab(TAB_INDEX_SEARCH);
     },
 
+    hasDuplicateWfsDownloadLinks: function(downloadHandlers) {
+
+        var wfsDownloads = [];
+        Ext.each(downloadHandlers, function(handler) {
+            Ext.each(handler.getDownloadOptions(), function(downloadOption) {
+                if (downloadOption.type == "WFS") {
+                    wfsDownloads.push(downloadOption.textKey)
```
Comment (jkburges) [inline web-app/js/portal/cart/DownloadPanel.js]: missing `;`

File: src/test/javascript/portal/prototypes/ArraySpec.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        expect([1,2,3,4,5].hasDuplicates()).toEqual(false);
+        expect([1,2,3,obj,4,5,obj].hasDuplicates()).toEqual(true);
+        expect([1,"string",3,obj,4,5,"string"].hasDuplicates()).toEqual(true);
+        expect([1,2,3,obj].hasDuplicates()).toEqual(false);
+    });
+
+});
```
Comment (jkburges) [inline src/test/javascript/portal/prototypes/ArraySpec.js]: missing end of line

File: web-app/js/portal/cart/DownloadPanel.js [hunk 5]
```diff
@@ -158,18 +158,37 @@ Portal.cart.DownloadPanel = Ext.extend(Ext.Panel, {
...
         Ext.each(downloadHandlers, function(handler) {
 
             Ext.each(handler.getDownloadOptions(), function(downloadOption) {
 
                 var newMenuItem = {
+                    name: handler.onlineResource.name,
+                    title: handler.onlineResource.title,
                     text: OpenLayers.i18n(downloadOption.textKey),
                     handler: function() {
                         this.confirmDownload(collection, this, downloadOption.handler, downloadOption.handlerParams, downloadOption.textKey)
```
Comment (jkburges) [inline web-app/js/portal/cart/DownloadPanel.js]: missing `;`

File: web-app/js/portal/cart/DownloadPanel.js [hunk 6]
```diff
@@ -158,18 +158,37 @@ Portal.cart.DownloadPanel = Ext.extend(Ext.Panel, {
...
+            Ext.each(handler.getDownloadOptions(), function(downloadOption) {
+                if (downloadOption.type == "WFS") {
+                    wfsDownloads.push(downloadOption.textKey)
+                }
+            });
+        });
+        return (wfsDownloads.hasDuplicates());
+    },
+
     _loadMenuItemsFromHandlers: function(processedValues, collection) {
```
Comment (jkburges) [inline web-app/js/portal/cart/DownloadPanel.js]: There's quite a lot of logic to do with creating the menu for each collection.  Consider extracting this out in to a separate class?

File: web-app/js/portal/cart/DownloadPanel.js [hunk 7]
```diff
@@ -158,18 +158,37 @@ Portal.cart.DownloadPanel = Ext.extend(Ext.Panel, {
...
 
         if (!processedValues.menuItems) {
             processedValues.menuItems = [];
         }
 
+        var groupedMenuItems = [];
+        var menuItemLabelsUsed = [];
         var downloadHandlers = Portal.cart.DownloadHandler.handlersForDataCollection(collection);
+        var hasDuplicateWFSDownloads = this.hasDuplicateWfsDownloadLinks(downloadHandlers);
+
```
Comment (jkburges) [inline web-app/js/portal/cart/DownloadPanel.js]: Can we extract a few functions out of this function, to make it a bit more readable? There are loops mixed with nested if/elses - pretty hard to understand currently.

File: web-app/js/portal/common/helpers.js [hunk 8]
```diff
@@ -324,3 +324,11 @@ function normaliseLongitude(longitude) {
...
     jQuery(jQuerySelector).empty();
 }
+
+Object.size = function(obj) {
+    var size = 0, key;
+    for (key in obj) {
+        if (obj.hasOwnProperty(key)) size++;
+    }
+    return size;
+};
```
Comment (jkburges) [inline web-app/js/portal/common/helpers.js]: missing end of line

File: web-app/js/portal/prototypes/Array.js [hunk 9]
```diff
@@ -11,3 +11,13 @@ Array.prototype.last = function() {
...
+
+//http://stackoverflow.com/questions/19655975/check-if-an-array-contains-duplicate-values
+Array.prototype.hasDuplicates = function() {
+    this.sort();
+    for ( var i = 1; i < this.length; i++ ){
+        if(this[i-1] == this[i])
+            return true;
+    }
+    return false;
+};
```
Comment (jkburges) [inline web-app/js/portal/prototypes/Array.js]: missing end of line

File: web-app/js/portal/prototypes/Array.js [hunk 10]
```diff
@@ -11,3 +11,13 @@ Array.prototype.last = function() {
         return this[this.length - 1]
     }
 };
+
+//http://stackoverflow.com/questions/19655975/check-if-an-array-contains-duplicate-values
+Array.prototype.hasDuplicates = function() {
+    this.sort();
+    for ( var i = 1; i < this.length; i++ ){
```
Comment (jkburges) [inline web-app/js/portal/prototypes/Array.js]: coding standard (spacing) - should be `for (var i = 1; i < this.length; i++) {`

File: web-app/js/portal/cart/DownloadPanel.js [hunk 11]
```diff
@@ -177,9 +196,42 @@ Portal.cart.DownloadPanel = Ext.extend(Ext.Panel, {
...
+            if (groupedMenuItems.hasOwnProperty(key)) {
+                processedValues.menuItems.push(OpenLayers.i18n('menuItemGroupTitle', {title: key}));
+                Ext.each(groupedMenuItems[key], function(item) {
+                    processedValues.menuItems.push(item);
+                });
+            }
+        };
+    },
+
+    getEmbededTitle: function(title) {
```
Comment (jkburges) [inline web-app/js/portal/cart/DownloadPanel.js]: spelling mistake - "embedded".  Sorry, I know that's pedantic, but it's annoying having to repeatedly "correct" something to the wrong spelling when you're using it. :-)

File: src/test/javascript/portal/cart/DownloadPanelSpec.js (hunk 1 above)
Comment (pmbohm) [inline src/test/javascript/portal/cart/DownloadPanelSpec.js]: Your right. Its a bit weird. For the moment we are limited as too what we can configure in Geonetwork. He's the last email on the subject from @pblain 

"Thanks Rog.  We will go with the brackets idea then.  
For everyone else's benefit, here's why I don't like it:

1. We will need to ask AODN contributors to conform to a rather peculiar standard if they want multi-WFS.  
2. If AODN contributors aren't aware of the "standard" they may innocently use brackets in their metadata for some other reason and get unexpected results
3. The long-winded description is redundant:  Users who know what a WFS is don't need it.  Users who don't know what a WFS won't be helped by it.
4. More people use the portal than GeoNetwork so the portal should be our priority.  
5. It's an inelegant and unwieldy solution - which means an increased chance of unanticipated problems.
Cheers,
Pete."

**Comment by pmbohm** [comment]
@jkburges New pull request pushed. "extract a few functions out of this function" - I think this is smaller now

**Pull request merged**